pkg/segment/segment_test.go

```diff
@@ -0,0 +1,191 @@
+package segment
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/preference"
+	"github.com/openshift/odo/pkg/version"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+type segmentResponse struct {
+	Batch []struct {
+		AnonymousID string `json:"anonymousId"`
+		MessageId   string `json:"messageId"`
+		Traits      struct {
+			OS string `json:"os"`
+		} `json:"traits"`
+		Properties struct {
+			Error     string `json:"error"`
+			ErrorType string `json:"error-type"`
+			Success   bool   `json:"success"`
+			Version   string `json:"version"`
+		} `json:"properties"`
+		Type string `json:"type"`
+	} `json:"batch"`
+	MessageID string `json:"messageId"`
+}
+
+func mockServer() (chan []byte, *httptest.Server) {
+	done := make(chan []byte, 1)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		bin, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		done <- bin
+	}))
+	return done, server
+}
+
+func TestClientUploadWithoutConsent(t *testing.T) {
+	body, server := mockServer()
+	defer server.Close()
+	defer close(body)
+	falseValue := false
+
+	cfg := &preference.PreferenceInfo{
+		Preference: preference.Preference{
+			OdoSettings: preference.OdoSettings{
+				ConsentTelemetry: &falseValue,
+			},
+		},
+	}
+
+	c, err := newCustomClient(cfg, filepath.Join(os.TempDir(), "telemetry"), server.URL)
+	//	assert err != nil
+	if err != nil {
+		t.Fatal(err)
+	}
+	// run a command, odo preference view
+	if err := c.Upload("odo preference view", time.Second, errors.New("an error occured")); err != nil {
```

Comment:
you need to mock the segment api responses as expected from segment and then validate the response instead of just checking for error?

Reply:
We are doing that in the other test `TestClientUploadWithConsent`. This test just makes sure that no data is received by the server if the user had not consented for telemetry.